designs/2019-config-simplification/README.md

```diff
@@ -0,0 +1,627 @@
+
+```js
+exports.config = (context) => {
+    const myConfig = {
+        files: ["*.js"],
+        ruledefs: {
+            custom: ruleLoader.loadFromDirectory("./custom-rules")
+        },
+        rules: {
+            "custom/my-rule": "error"
+        }
+    };
+
+    if (context.hasRule("some-new-rule")) {
+        myConfig.rules["some-new-rule"] = ["error"];
+    }
+
+    return myConfig;
+};
+```
+
+### Configuration Location Resolution
+
+When ESLint is executed, the following steps are taken to find the `eslint.config.js` file to use:
+
+1. If the `-c` flag is used then the specified configuration file is used. There is no further search performed.
+1. Otherwise:
+    a. Look for `eslint.config.js` in the current working directory. If found, stop searching and use that file.
+    b. If not found, search up the directory hierarchy looking for `eslint.config.js`.
+    c. If a `eslint.config.js` file is found at any point, stop searching and use that file.
+    d. If `/` is reached without finding `eslint.config.js`, then stop searching and output a "no configuration found" error.
+
+This approach will allow running ESLint from within a subdirectory of a project and get the same result as when ESLint is run from the project's root directory (the one where `eslint.config.js` is found).
+
+Some of the key differences from the way ESLint's configuration resolution works today are:
+
+1. There is no automatic search for `eslint.config.js` in the user's home directory. Users wanting this functionality can either pass a home directory file using `-c` or manually read in that file from their `eslint.config.js` file.
+1. Once a `eslint.config.js` file is found, there is no more searching for any further config files.
+1. There is no automatic merging of config files using either `extends` or `overrides`.
+1. When `-c` is passed on the command line, there is no search performed.
+
+### File Pattern Resolution
+
+Because there are file patterns included in `eslint.config.js`, this requires a change to how ESLint decides which files to lint. The process for determining which files to lint is:
+
+1. When a filename is passed directly (such as `eslint foo.js`):
+    1. ESLint checks to see if there is one or more configs where the `files` pattern matches the file that was passed in and does not match the `ignores` pattern. The pattern is evaluated by prepending the directory in which `eslint.config.js` was found to each pattern in `files`. All configs that match `files` and not `ignores` are merged (with the last matching config taking precedence over others). If no config is found, then the file is ignored with an appropriate warning.
```

Comment:
Could you elaborate on how `files` and `ignores` are merged with `extends`? 

```js
exports.config = {
    files: ["*.js", "*.ts"],
    extends: [
        {
            files: ["*.ts"]
            ...extendedConfig
        }
    ],
    ...baseConfig
};
```

If I'm understanding the discussion at https://github.com/eslint/rfcs/pull/9#discussion_r251180712 correctly, would that config be equivalent to this?

```js
exports.config = [
    {
        files: ["*.ts"]
        ...extendedConfig
    },
    {
        files: ["*.js", "*.ts"],
        ...baseConfig
    }
];
```

Reply:
This is one of the open questions I listed in the design. My initial thought was that `files` and `ignores` would be overwritten rather than merged.

However, if we remove `extends` (as @mysticatea has suggested), then I think this becomes a lot clearer.

Reply:
Another option would be to make `extends` be syntax sugar for array configs, i.e.

```js
module.exports = {
    ...config,
    extends: [shareableConfig]
}
```

...would always be equivalent to:

```js
module.exports = [
    shareableConfig,
    config
]
```

In most cases, I think shareable configs probably wouldn't specify file patterns anyway, and having `extends` might be more familiar/readable to new users (who otherwise might not be sure what array configs do). I don't feel strongly about this, however.

Reply:
Please take a look at the latest update. I think removing `extends` actually makes things a lot simpler. Basically, a config array acts just like an `extends` array does now, so I think it will feel pretty familiar to users.